### DISCUSSION TITLE
Error when loading the parser without vocab (working in 2.1.9, stopped in 3.1.0)

### DISCUSSION THREAD

**Discussion by oliviercwa** [explosion/spaCy#8939] [Help: Coding & Implementations] [answered]
This scenario is working fine in 2.1.9 but fails in 3.1.1.

Our goal is to load separately the vocab from other components as we package them both separately on our system to speed up the pipeline loading process. In 3.1.1 when loading the parser, an exception is thrown

Any suggestion to work around the issue ?

Thank you

## How to reproduce the behaviour
```python
  model_path = Path(os.path.join(distutils.sysconfig.get_python_lib(), 'en_core_web_md', 'en_core_web_md-3.1.0'))
  # Exclude all components
  pipeline = spacy.load(model_path, exclude=["tok2vec", "tagger", "parser", "attribute_ruler", "lemmatizer", "ner"])

  # build pipeline starting with tokenizer
  tokenizer = spacy.tokenizer.Tokenizer(pipeline.vocab)
  tokenizer.from_disk(model_path / 'tokenizer')
  pipeline.tokenizer = tokenizer

  # Load parser
  parser = spacy.pipeline.DependencyParser(pipeline.vocab, DEFAULT_PARSER_MODEL)
  parser.from_disk(model_path / 'parser', exclude=['vocab'])

  pipeline.add_pipe('parser')
```
I have also tried to load the config separately without any luck

```
  ...

  # Load parser
  config: Config = Config()
  config.update(srsly.read_json(model_path / 'parser' / 'cfg'))
  parser = spacy.pipeline.DependencyParser(pipeline.vocab, config)
  parser.from_disk(model_path / 'parser', exclude=['vocab'])

  pipeline.add_pipe('parser')
```

## Result
```Exception has occurred: ValueError
[E149] Error deserializing model. Check that the config used to create the component matches the model being loaded.
  File "spaCy\spacy\pipeline\transition_parser.pyx", line 591, in spacy.pipeline.transition_parser.Parser.from_disk
    self._resize()
  File "spaCy\spacy\pipeline\transition_parser.pyx", line 189, in spacy.pipeline.transition_parser.Parser._resize
    self.model.attrs["resize_output"](self.model, self.moves.n_moves)

During handling of the above exception, another exception occurred:

  File "spaCy\spacy\pipeline\transition_parser.pyx", line 594, in spacy.pipeline.transition_parser.Parser.from_disk
    raise ValueError(Errors.E149)
  File "tests\sentencizer_test.py", line 27, in test_spacy
    parser.from_disk(model_path / 'parser', exclude=['vocab'])
  File "test\sentencizer_test.py", line 98, in <module>  test_spacy()
```

## Your Environment

* Operating System: Win 10
* Python Version Used: 3.7.4
* spaCy Version Used: 3.1.1
* Environment Information:

**Comment by adrianeboyd** [comment, marked as answer]
You should be able to do this:

```python
vocab = load_my_vocab()
nlp = spacy.load("model_package_name", exclude=[...everything except the parser...], vocab=vocab)
```

There are quite a few of details under the hood here in order to load and initialize the pipeline components from the config. If you want to do it by hand, follow the steps here:

https://github.com/explosion/spaCy/blob/e227d24d4356ddd6ffc0b997a092c02225bbc3e3/spacy/util.py#L360-L393

You have to:

* load the config
* initialize the pipeline from the config
* load the components from disk (model weights, lemmatizer tables, etc.)

If you're initializing the parser separately, the config has to match the config from `[components.parser]` in `config.cfg`, which is neither `DEFAULT_PARSER_CONFIG` (which is a valid config for the component, just not the right one) nor the settings in `parser/cfg` (which is only a subset of settings and not a full component config).

In v3, if you instantiate a parser like this, you can't actually add it to a pipeline later because you can't recover the necessary config details after the internal model is instantiated:

```python
parser = spacy.pipeline.DependencyParser(pipeline.vocab, args...)
```

Instead, always use `add_pipe`:

```python
nlp.add_pipe("parser", config=...)
```
